### DISCUSSION TITLE
apple silicon m2pro & m4 速测新的macvk后端

### DISCUSSION THREAD

**Discussion by hooke007** [hooke007/mpv_PlayKit#376] [mpv 通用]
> TL;DR
> 多年后，mac的mpv用户终于能赶上win/Linux用户的使用体验了。。。
> 这块GPU是"垃圾"，无法满足高阶着色器的需求

此节点起 https://github.com/mpv-player/mpv/commit/78d43740f52db817d98bcf24fb30a76ab6fa13ff mpv for MAC 正式可启用 vulkan 后端。
需要一个足够新的版本才可使用，自行编译步骤可参考 https://github.com/hooke007/mpv_PlayKit/discussions/70
如何使用macvk的配置可参考上文的二层

<img width="600" alt="截屏2023-10-17 21 01 45" src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/1e15291b-ba35-446f-944a-591f3bc8cd2a">

##
新后端的引入意味着终于可以用上主流的vo了，功能不再严重缺失。
首先是简单的4k60片源的软解，负载比较如图（左 vo=libmpv 右 vo=gpu）
图中没体现出来的是，macvk没有肉眼可见的卡帧现象。

<img width="200" alt="负载 libmpv og" src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/75456d6d-1491-42ce-b9d5-f4b6c41e767d"> <img width="200" alt="负载 gpu vk" src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/5f395ddc-3e63-4188-a788-28a4c7a06923">

##
mac自带的color sync会影响mpv的色彩表现（左 默认icc 右 srgb icc）
启用icc-profile-auto的时候请谨慎。

<img width="300" alt="色彩管理 默认icc" src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/f3d991ab-6955-482f-88c8-c97db790aba0"> <img width="300" alt="色彩管理 srgb icc" src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/0b99aa44-6850-4793-ba55-1f2927e7c208">

##
因为跟上了版本，因此缺失的部分功能比如动态色调映射可以正常工作，解决hdr转sdr过曝的问题。
（但是mbp的这块内屏直出hdr的表现已足够，mbp的内屏用户基本不需要sdr映射功能）

<img width="300" alt="hdr 静态元数据" src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/a32bef79-bad8-4666-b5a9-9064242b2a99"> <img width="300" alt="hdr 动态测光" src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/782ad421-9194-47bb-b76d-aedb42bfcf28">

##
应用着色器后的性能比我想象的要弱得多 `Anime4K_Clamp_Highlights.glsl:Anime4K_Restore_CNN_VL.glsl:Anime4K_Upscale_CNN_x2_VL.glsl` 这个组合无法正常使用。
最终在vo=gpu-next下对常规的1080p24采用了 `Anime4K_Restore_CNN_L.glsl:Anime4K_Upscale_CNN_x2_L.glsl` 显卡负载基本稳定90%左右（这个爆炸的帧计时看着直摇头）

<img width="600" alt="截屏2023-10-18 19 34 54" src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/8afd2048-071b-4753-a7e5-f0934d0a514f">

> 也测试了anime4kmetal的性能（使用quicktime解码渲染）使用同前的组合项，图形负载仅低了5%，看来是真的不行。
> 
> <img width="400" alt="截屏2023-10-18 22 00 23" src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/905dc16d-29d2-4bda-9c65-29f5f7d13c8e">

##
其它：
- 由a4k的测试结果可见高阶着色器（主要是组合或需要多次应用的场景中）的使用存在性能制约，其它着色器的快速测试结果
 nnedi3-64-8x6 掉帧严重，只能降级到8x4；fsrcnnx-16 单开通过，图形负载约80%+；ravu-zoom-r3-rgb 单开通过，负载约30%左右
- 当前macvk无法激活native硬解码，会自动跳转copy。 https://github.com/mpv-player/mpv/issues/12675
 vulkan解码也不可用。
- 可以直通hdr，但是退出时会产生段错误。 https://github.com/mpv-player/mpv/issues/12666
- 十分费劲的部署上了 rife-ncnn-vk 但是连常规的720p也很难做到始终流畅倍帧。

**Comment by eko5624** [comment]
在我的macbook pro 2017测试性能表现比`gpu-context=auto`好一丢丢。
<img width="1440" alt="QQ20231017-221202@2x" src="https://github.com/hooke007/MPV_lazy/assets/84926938/ffdca5c2-d7c7-4cb0-9281-84756e899940">

**Comment by eko5624** [comment]
或许实现native硬解码后性能表现会好点吧。

**Reply by hooke007** [reply]
对这块m2p来说cpu多吃十几个百分比倒是无所谓。图形负载占的太多，费电明显且给想要折腾画面的性能余量太少。

而第一个例子，同样的片源里其他metal播放器的性能表现是这样的
<img width="200" alt="截屏2023-10-17 21 19 19" src="https://github.com/hooke007/mpv_PlayKit/assets/41094733/909c398e-4ffd-466f-8e5f-fff6b3e35595">

**Comment by eko5624** [comment]
打了这个补丁后可以启用HW videotoolbox了。不过对性能没有太大改善。
https://github.com/m154k1/mpv-build-macOS/blob/master/patches/mpv/0001-vo-gpu-next-videotoolbox.patch

**Comment by eko5624** [comment]
神奇的事情发生了，这个提交后 `vo=gpu-next gpu-context=macvk` 性能变好了很多。上面的测试视频居然不掉帧了。。
https://github.com/mpv-player/mpv/commit/bd1ba99ab2f64f218128100cbf3a91e5cb1b4872

**Reply by hooke007** [reply]
intel那边应该和vd-lavc-dr有关联

**Comment by hooke007** [comment]
<img width="300" src="https://github.com/user-attachments/assets/4cb92b99-3b73-45aa-ad30-715175a2d702" />

2024年发布的Mac mini m4，跑个残废的nnedi3 32已经是极限了，仍然只是个大号的核显。。。

<img width="500" src="https://github.com/user-attachments/assets/00f10986-e0cb-4511-8d67-1309e21d19d1">
